migration/migrator/migrations/master/20250121144600_make_empty_preferred_names_null.py

```diff
@@ -0,0 +1,33 @@
+"""Migration for the Submitty master database."""
+
+
+def up(config, database):
+    """
+    Run up migration.
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    """
+
+    database.execute("UPDATE users SET user_preferred_givenname = NULL WHERE user_preferred_givenname = ''")
+    database.execute("UPDATE users SET user_preferred_familyname = NULL WHERE user_preferred_familyname = ''")
+    database.execute("ALTER TABLE users ADD CONSTRAINT user_preferred_givenname_not_empty CHECK (user_preferred_givenname <> '')")
+    database.execute("ALTER TABLE users ADD CONSTRAINT user_preferred_familyname_not_empty CHECK (user_preferred_familyname <> '')")
+    pass
+
+
+def down(config, database):
+    """
+    Run down migration (rollback).
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    """
+
+    database.execute("ALTER TABLE users DROP CONSTRAINT user_preferred_givenname_not_empty")
```

Comment:
I'm assuming we would also want to covert all nulls back to empty string as that is what the PHP code would expect?

Reply:
No, the php code is already built to handle both, the goal of this PR is to start the transition to having only one empty value instead of two

Reply:
Where can the `null` values get introduced into the database for this column currently?

Reply:
When developing locally all the sample data is currently null, everywhere the value is used the database statement looks like `COALESCE(NULLIFpreferred_name),''),legalname)`

Reply:
Sounds like there's a mistake in the sample courses script. However, it will be fine after this goes in.

> everywhere the value is used the database statement looks like `COALESCE(NULLIF(preferred_name),''),legalname)`

The user model scoops up preferred name without this condition. That gets used all over the place in Submitty. Even with the `coalesce(nullif(...), ...)` it still seems the underlying data is always an empty string rather than null (at least for production). I'm not seeing any other way in the code for null values to get into the database. So in favor of accuracy, it would probably be best to revert all nulls to empty string in the down migration.